### PR TITLE
HPCC-28587 Add support for building with AddressSanitizer

### PR DESCRIPTION
**Pull request by jakesmith**
with suppression of leak checking for problematic areas:
+ esdl-xml
+ hidl
+ xalan leak as used by configen.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/scm/additional.cmake [hunk 1]
```diff
@@ -39,11 +39,11 @@ foreach ( loop_var ${ESPSCM_SRCS} )
     if (SCM_BUILD)
       add_custom_command ( DEPENDS hidl ${ESPSCM_SOURCE_DIR}/${loop_var}
                            OUTPUT ${ESPSCM_GENERATED_DIR}/${result}.esp ${ESPSCM_GENERATED_DIR}/${result}.hpp ${ESPSCM_GENERATED_DIR}/${result}.int ${ESPSCM_GENERATED_DIR}/${result}.ipp ${ESPSCM_GENERATED_DIR}/${result}_esp.cpp ${ESPSCM_GENERATED_DIR}/${result}_esp.ipp
-                           COMMAND $<TARGET_FILE:hidl> ${ESPSCM_SOURCE_DIR}/${result}.ecm ${ESPSCM_GENERATED_DIR}
+                           COMMAND ASAN_OPTIONS=detect_leaks=0 $<TARGET_FILE:hidl> ${ESPSCM_SOURCE_DIR}/${result}.ecm ${ESPSCM_GENERATED_DIR}
```
Comment (jakesmith) [inline esp/scm/additional.cmake:42]: NB: I have opened https://track.hpccsystems.com/browse/HPCC-28588 to pursue the leaks caused by hidl and esdl-xml, but those fixes are non-trivial and can be dealt with separately.

File: initfiles/bash/etc/init.d/hpcc_common.in [hunk 2]
```diff
@@ -313,6 +313,7 @@ configGenCmd() {
     # Creating logfiles for component
     logDir=$log/${compName}
 
+    export LSAN_OPTIONS=suppressions=${path}/etc/init.d/lsan-supp.txt
```
Comment (jakesmith) [inline initfiles/bash/etc/init.d/hpcc_common.in:316]: This _could_ go ahead of other component uses, but there wasn't any other uses cases at the moment. We can move it (also add it) to other places if deemeed necessary later.

But this is specifically to address a leak in xmlStrndup in xalan as used by configgen, which I did not look like configgen's fault, but the 3rd party library - possibly would be fixed by using a more modern version

**Review by richardkchapman — APPROVED** [review]

**Comment by mckellyln** [comment]
Great PR.  I have also had to suppress huge hidl/esp output.
I think I also tried ASAN_OPTIONS=fast_unwind_on_malloc=0 if trace info was missing.
I have also used -fsanitize=leak option and then with that LSAN_OPTIONS="detect_leaks=0"

**Review by mckellyln — APPROVED** [review]
Approved

**Review by ghalliday — APPROVED** [review]

File: esp/scm/additional.cmake (hunk 1 above)
Comment (ghalliday) [inline esp/scm/additional.cmake:42]: @jakesmith does this work for windows compiles?

**Review by ghalliday — CHANGES_REQUESTED** [review]
Pending windows check/change

**Comment by jakesmith** [comment]
@ghalliday - added change to avoid cli asan option usage in Windows.
Let's wait for github windows build to run 1st though.

**Comment by jakesmith** [comment]
> @ghalliday - added change to avoid cli asan option usage in Windows. Let's wait for github windows build to run 1st though.

@ghalliday - it successfully completed the windows build.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**